fenix/app/src/main/java/org/mozilla/fenix/compose/ThumbnailImage.kt

```diff
@@ -0,0 +1,106 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.compose
+
+import android.graphics.Bitmap
+import androidx.compose.foundation.Image
+import androidx.compose.foundation.background
+import androidx.compose.foundation.layout.Box
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.rememberCoroutineScope
+import androidx.compose.runtime.saveable.rememberSaveable
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.asImageBitmap
+import androidx.compose.ui.graphics.painter.BitmapPainter
+import androidx.compose.ui.layout.ContentScale
+import androidx.compose.ui.platform.LocalDensity
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.dp
+import kotlinx.coroutines.launch
+import mozilla.components.concept.base.images.ImageLoadRequest
+import org.mozilla.fenix.components.components
+import org.mozilla.fenix.theme.FirefoxTheme
+
+/**
+ * Thumbnail belonging to a [key]. Asynchronously fetches the bitmap from storage.
+ *
+ * @param key Key used to remember the thumbnail for future compositions.
+ * @param size [Dp] size of the thumbnail.
+ * @param modifier [Modifier] used to draw the image content.
+ * @param contentScale [ContentScale] used to draw image content.
+ * @param alignment [Alignment] used to draw the image content.
+ */
+@Composable
+@Suppress("LongParameterList")
+fun ThumbnailImage(
+    key: String,
+    size: Dp,
+    modifier: Modifier,
+    contentScale: ContentScale,
+    alignment: Alignment,
+    fallbackContent: @Composable () -> Unit,
+) {
+    if (inComposePreview) {
+        Box(modifier = Modifier.background(color = FirefoxTheme.colors.layer3))
+    } else {
+        val thumbnailSize = LocalDensity.current.run { size.toPx().toInt() }
+        val request = ImageLoadRequest(key, thumbnailSize)
+        val storage = components.core.thumbnailStorage
+        var state by rememberSaveable { mutableStateOf<Pair<Bitmap?, Boolean>>(Pair(null, false)) }
```

Comment:
optional nit: I prefer private data classes to `Pair`s, because it adds more semantic value to the type. That is, it requires more attention to remember what `.first` and `.second` are compared to `.bitmap` and `.hasLoaded`. It might also save the re-declaration  of `val (bitmap, hasLoaded) = state`